### ISSUE TITLE
[2018] Neural Ordinary Differential Equations

### ISSUE THREAD

**Issue by yoheikikuta** [yoheikikuta/paper-reading#21]
## 論文リンク
https://arxiv.org/abs/1806.07366

## 公開日（yyyy/mm/dd）
2018/01/19

## 概要
TBW.

**Comment by yoheikikuta** [comment]
そろそろ NeurlPS2018 の論文を読んでみようと思って、色々話題になっていそうなこの論文から読んでみることにした。
ちゃんとチェックしてないが、解説記事が出てたり [reddit](https://www.reddit.com/r/MachineLearning/comments/a65v5r/neural_ordinary_differential_equations_pdf/) もなかなか盛り上がっていそうだ。こういうの見ても結局理解はできないので自分で読んでみる。

**Comment by yoheikikuta** [comment]
著者による公式実装はこれ：https://github.com/rtqichen/torchdiffeq
PyTorch 実装。

中身は必要に迫られたら読むことにする。

**Comment by yoheikikuta** [comment]
まずは問題のセットアップと何をしたいのかというところから。

ResNet や RNN の hidden state の構造は次のように書くことができる。
<p align="center">
  <img src="https://i.imgur.com/GU7hmQ4.png">
</p>

これを $` t \rightarrow t+1 `$ という刻み幅での更新式だと考えれば、layer の数を増やして刻み幅が 0 となるような極限を取ることで次のような常微分方程式とみなすことができる。

<p align="center">
  <img src="https://i.imgur.com/uZ4pPXb.png">
</p>

初期条件 h(0) を与えてこの ODE を解けば、任意の時間 T での h(T) が求まる。
このようにある種の Neural Network を常微分方程式と考えることはこれまでもされてきたことだが、この論文ではこの ODE を adjoint method という時間を逆方向に辿って解く手法（これは通常の back propagation）で実際に ODE を解くことでネットワークを構築する。

こうすると何がいいのかということは以降で見ていく。

**Comment by yoheikikuta** [comment]
連続化することによるメリットは以下の通り。

- Memory efficiency
  通常の back propagation で必要な中間 node の値を全て保持しておくという必要がなく、ODE を解くことでネットワークを学習していけるので、その意味（離散の場合は layer を追加するごとに必要なメモリが大きくなるがそれがない）でメモリ効率が良い。本当は ODE をどう解くかに依存する話だが、少なくとも連続化によって layer という概念はなくなっている。
- Adaptive computation
  ODE を解く際には近似誤差をコントロールすることができる。例えばオイラー法とルンゲクッタ法では解くスピードや精度に違いが出てくる。つまり解きたい問題の複雑性や投入できる計算リソースに合わせて ODE の solver を選択できるということだ。
- Parameter efficiency
  離散の場合は layer を追加するごとにパラメタ数が増えていくが、連続化して layer 数が無限個の極限で一つの ODE を構成するため、その ODE を特徴付けるパラメタだけで済むためパラメタ数を抑えられる可能性がある。
- Scalable and invertible normalizing flows 
  副産物として、normalizing flow に代表される変数変換則が簡単に解けるようになる。normalizing flow では変数変換を繰り返して複雑な事後分布を近似できるようにしていくが、その過程にも同じく連続化手法を適用する。これは normalizing flow に詳しくないのでちょっと把握できてない。
- Continuous time-series models
  離散化をベースとする RNN とは違い連続的な時系列モデルを組むことができる。

実際にこれらがどれくらい嬉しいのかを詳しく見ていく。

**Comment by yoheikikuta** [comment]
ODE として解くというのはどういうことなのか、真面目に数式を追ってみる。
連続化した中間層から最終的な出力と loss を計算するには、以下のように書ける。ここで時間積分が ODE を解くことに相当するので、そこを典型的に ODESolve と書いている。
ここで ODESolve は任意の常微分方程式の solver で blackbox（ここでの意味はネットワークとは切り離されていて別個に解くことができるもの）である。

<p align="center">
  <img src="https://i.imgur.com/CyxKPYc.png">
</p>

最終的には loss が小さくなるようなモデルを構築したいので、各パラメタ（初期値の z(t_0), t_0 と終端の t_1 とパラメタ θ）での微分値を求めて backpropagation で学習をしていく必要がある。

単純に forward pass で解いていくのは、普通の back propagation を forward mode で解く場合と同様にパラメタ数が多いと大変（という理解）。

こいつも reverse mode に相当する解き方で解くことができる。それが adjoint method である。
基本的な量としては以下の adjoint である。

<p align="center">
  <img src="https://i.imgur.com/ySTCcdK.png">
</p>

この adjoint の dynamics を解くことで必要な微分を計算する。
初期時刻での微分値は ∂L/∂z(t_0) は以下の常微分方程式を解けば良い。ただしこれは z(t_1) という終端時刻の値を初期値として時間の逆方向に解くことに注意する。正に reverse mode になっているというわけだ。

<p align="center">
  <img src="https://i.imgur.com/sd3P1qo.png">
</p>

a がこの常微分方程式を満足することは自明でなく、Appendix B にて証明されている。
基本的には時間に対する微小変化 ε を考えてその Taylor 展開を真面目に計算するという程度なので follow するのは難しくない。

これを解くためには各時刻での z(t) の値を知らないといけないが、論文には以下のように adjoint と一緒に計算していけると書いてある。これは最初に L を計算する際に得られるものだと思うのだが、そういうものではないのだろうか？別の ODE だから得られる値が別になるのか。

```
One complication is that solving this ODE requires the knowing value of z(t) along its entire trajectory. 
However, we can simply recompute z(t) backwards in time together with the adjoint, starting from its final value z(t_1).
```

この辺りの事情を図示したのものが下図。
Adjoint state の各 step で値が update されているのは、a の常微分方程式を積分して解いた時の終端の値（a(t_i+1) - a(t_i) の二項目）の効果になっている。

<p align="center">
  <img src="https://i.imgur.com/NTP0jSn.png">
</p>

この辺が肝で、あとは同様に ∂L/∂θ などを求めていく。

**Comment by yoheikikuta** [comment]
θ や t での微分値に関しても同様に求められて、それは Appendix に記載されている。
この論文は Appendix の説明が丁寧なので式を追うこと自体は難しくない。スカラーなのかベクトルなのかはちょっと分かりづらいけど。f とかベクトルなので bold で書いて欲しいけどね。

なんにせよ、これで学習に必要な微分値が求められ、以下のようにまとめられる。
ここでは一度 forward-time ODE で L は求めた後の話になっている。
そして既出だが、ODE solver は任意の solver でネットワークとは別に解くというものになっている。この論文での実験では Python autograd の coe から fortran の ODE を呼んでると言っている。

<p align="center">
  <img src="https://i.imgur.com/ECHk0hO.png">
</p>

準備はできたので、後はこのモデルで問題を解いた時に本当に既存のものよりも良い結果を出してくれるのかを調べる。

**Comment by yoheikikuta** [comment]
MNIST で比較した結果が以下。
MLP はパラメタ数が同じくらいになるように、ResNet は性能が同じくらいになるように（residual block が 6 つになる）して、提案手法と比較している感じ。

ここで、L は layer 数で $` \tilde{L} `$ は ODE solver で一回の forward pass において必要となる関数の評価回数となっている。
これは ODE を特徴付けるもので、実装では [こんな感じ](https://github.com/rtqichen/torchdiffeq/blob/8df757cb12f231a6b4349a96608b7a9d11166988/examples/odenet_mnist.py#L92-L113) のようになっている。6 個 residual block を重ねるのと比べるとだいぶ少ないこと分かる。

ちなみに ODE と ResNet を入れ替えてる部分は実装では [ここ](https://github.com/rtqichen/torchdiffeq/blob/8df757cb12f231a6b4349a96608b7a9d11166988/examples/odenet_mnist.py#L304) に現れている。ODE の方は一発だけ block を入れていて、ResNet は 6 個繰り返していることが分かる。
と思ったけどこれはさらに刻み幅の分だけ処理が走るわけだから、その積になるということかな。精度との tradeoff になるのはいいとして、適切に選べば既存のものよりも本当に速いのか？というのはよく分からない。

<p align="center">
  <img src="https://i.imgur.com/vV1LgDB.png">
</p>

ちなみに ODE-Net がアダムス法で ODE を解くもので、RK-Net がルンゲクッタ法で ODE を解くものになっている。学部生のころやった気がするけど覚えてないね。ルンゲクッタがテイラー展開の高次まで取ったものだった気がするけど、アダムス法は分からん。

まあともかく結果としては少ないパラメタで必要なメモリも少なく、そして良い性能を発揮しているということが分かる。分かるけど、MNIST かぁという感じはする。まあ話題になっているのでもっと色々な問題に適用されてテストされるでしょう。

**Comment by yoheikikuta** [comment]
次に ODE のエラーコントロールの話。
例えば一番左の図は、自分の理解では、縦軸が数値計算の誤差で横軸が forward pass における関数評価の回数で色が tolerance になっている。なので tolerance を低くするのはすなわち刻み幅を細かくすることに対応していて、それゆえに誤差が小さくなるということになっている。
tolerance を低くしても関数評価の数が 100 とかそれくらいなので、スピードも十分出るのかな。速度の定量的な比較も欲しいところだ。

<p align="center">
  <img src="https://i.imgur.com/vDGuIwe.png">
</p>

(c) の図がなかなか面白くて、backward の関数評価回数は forward の関数評価回数の半分くらいで、メモリ効率が良いだけでなくて計算効率も良いということになっている。
でもこれは何でだろう？理由はちょっと分からない。ODE-Net がどうこうという話よりも ODE を adjoint method で解く際の話に思えるので調べられていそうなものだが。

まあ何にせよ、これで ODE のエラーコントロールが機能していることが分かる。
学習時には tolerance を低く設定してパラメタを更新して、テスト時には tolerance を高くして誤差を犠牲にして早く動かす、みたいなこともできる。

一個気を付けたいのはここでの誤差はあくまで数値計算上の誤差でしかなくて、tolerance を小さくすれば大域解にたどり着くとかそういう類のものではない。

**Comment by yoheikikuta** [comment]
ここまでが論文のメインの内容で、次は normalizing flow を連続化して効率的に求められるという話。

この辺りは [Glow](https://arxiv.org/abs/1807.03039) が出た辺りでちゃんと追っていればもっと楽しめたと思うが、あまり追ってないのでよく分からない。

自分の理解では、正規分布のような単純な確率分布の潜在変数に対して、可逆な関数を適用していって複雑な分布、すなわちデータ空間の確率分布を表現するというものになっている。

この可逆な関数の適用を一回だけ抜き出すと以下のようになっている。
対数尤度最大化が目的になるのでそちらの式も書いている。

<p align="center">
  <img src="https://i.imgur.com/vUuWTDj.png">
</p>

計算としては Jacobian の det が大変だが、これを連続化した式にすることで効率的に計算できるというのがこの論文で示されている。

<p align="center">
  <img src="https://i.imgur.com/d5F1Ttq.png">
</p>

これは trace だけなので計算が楽。証明は Appendix にあるが、丁寧に式変形を書いてくれてるので読めば分かる。

ちょっとここからは省略するが、これを用いて真の確率密度とどれくらいマッチするかを通常の Normalizing Flow (NF) と比べたのが以下の図。
提案手法（CNF）の方が loss が小さくなっている。なぜか性能の話にすり替わっていてしかも性能が上がるのは何でなのかはちょっとよくわかってない。NF の理解も合わせて誰か詳しい人と議論したいなぁ。

<p align="center">
  <img src="https://i.imgur.com/XroRE0B.png">
</p>

**Comment by yoheikikuta** [comment]
最後に時系列モデルを連続化して解くという話。
ここでは潜在変数の時間発展を ODE で解くことで、データ欠損などに対応できるモデルを作りたいというモチベーションだ。そういう話は自分は詳しくないが前からやられているようで、この論文で提案した枠組みでそれもできるぞ、ということになっている。

全体の構造としては以下。

<p align="center">
  <img src="https://i.imgur.com/kjEzaw9.png">
</p>

通常の離散化して解くケースと比べて、滑らかな分布に基づくものならば連続化がうまくいくような気はする。

実際に、よくやる実験で、時間順序にしたがって二つの spiral 状のデータ点を生成し、それをランダムにサンプルして分布を学習して test することで、特に学習データ点が少ない場合の RMSE で優秀な成績を示している。

また、spiral 状の分布のデータから潜在変数を学習し、それに基づいて以降のデータ点を補間する、という実験でも以下のように綺麗に再現をした。なかなかインパクトのある絵だ。
まあこれはきれいで滑らかな分布から生成されているので提案手法が良いのは reasonable だとは思う。

<p align="center">
  <img src="https://i.imgur.com/MFnH4Mm.png">
</p>

**Comment by yoheikikuta** [comment]
その他のトピックとして以下のようなものを軽く議論している。
- mini batch の扱いが通常の NN と比べて複雑（ただし実用上はそこまで問題になってないと主張）
- 微分方程式がリプシッツ条件を満足すれば解の存在と一意性を保証してくれる
- ODE solver の誤差や異なる複数の初期条件から同じ最終状態に写そうとすることによる情報量の喪失、によって可逆性が崩れる（あんまりちゃんと理解してないけど実用上はそこまで問題じゃないと主張）

**Comment by yoheikikuta** [comment]
ところどころ怪しいところはあるけど、骨子の部分は理解できたと思う。
話題になってる論文のようだし、きっとこの先読んだ人と遭遇して議論することで理解を深められるでしょう（希望的観測）。
